### PR TITLE
Make root namespace optional

### PR DESCRIPTION
**Pull request by Frando**
Hi,

this PR makes the root namespace in the generated typescript file optional. This module works just fine without using namespaces when exporting different collections of types to different files. Other typescript code can import these as ESM modules.

As the members of the option struct are public, this is a breaking change which would nee a semver bump to `0.4`.

Thanks!

### CODE REVIEW

**Comment by dbeckwith** [comment]
I remember the reason I added the root namespace in the first place was that without I was having issues with type name ambiguity in the generated Typescript in some cases. Let me test that out again and post an example if I can. If I'm right, I might still accept this change but with a note in the documentation giving an example of edge cases where this can go wrong.

**Comment by dbeckwith** [comment]
[Here's an example of what I was talking about](https://www.typescriptlang.org/play?#code/PTAEBUAsEsGdQPYFcAuBTATqFBPADmqJAIbwB2CoGCCKoZxAtmrHsQMZoA0osl0ddsTKIyAGxygARoQxoAZpjkATbJVKgAVklh0ABgDEaegHQBYAFBoAHngQY6uAqCOUAvKADeoawC5eKBjQZADmANygAL5hljZ2DvRMLGycoMFkmF6WoDmgIBAw8OmZToRwaWSw0MqExInMrBzcvPwo8EIUZNBCYvVl8HrFGCauptm5cfaO+ISuoB7eOP66QaER0eM5+VDlQ9gzafCBOMEhalQKJZQokITI6FgIGVybeWBSqNi3oEhkAI5IYhiaDyaBoVQMZigQzGQ4VXRoYiqYIPPBydCqDRyRRyMjsU6vfI3QhiBA9fbOGEIPTw9BIxDyL6yGgoAC0YjQADc0L0nmhzBZcqBJglSqAAELELALUDyGj+VzrGIWSKWSz5ACC8DqclgSDEKB4N3KIWg3O1FMISns0jQQh0hD0kuGcupRA0YsGZAyw1GPAo+lGAo5dCkUv8QxMzvmXll8tjflAAHIk1Eosr1WAAJKMoG8xnEy3auQ-Mg1LA6ll9RqcHhczCSMUkbUiX5MKTQEJIZDweT6iSsgFAkFgiFJAUiuiQ5JNItZQUTWxTS0uGgx7yJlanJWvSfVlJlb2ZTyvIV7sVzGVLAKrcLp0+5baFCo+lcdH6wQiFtuMDtdnt9NCpSwCMsIoJQ2JXEyiCoJkfICkKZ5LqKBzRjKrr+MBoEIDuC45KqKpqhYT7lAA7vYADW2rwHEdoYsGaChlKABMmEzCBkZobGGEJv4KZptEQA). The cases where this happens are pretty specific and might not come up that often, so I think I'm okay with just adding a comment warning about it to the documentation. I can do that myself after this is merged.

File: src/impls.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ impl_native!(Blob, "Blob");
 impl_native!(bool, "boolean");
 impl_native!(String, "string");
 impl_native!(str, "string");
+impl_native!(&'static str, "string");
```
Comment (dbeckwith) [inline src/impls.rs:36]: Given that there's already a `impl_native!(str, "string");` and a `impl<T: TypeDef> TypeDef for &'static T` in this file, I would expect this not to be needed. Did you find a situation that required this new impl?

**Review by dbeckwith — CHANGES_REQUESTED** [review]

**Comment by Frando** [comment]
Hi, thanks for the review and the example of how using nested namespaces without a root namespace fails. My usecase is currently not using any nested namespaces, I see the issue with using nested namespaces without a root namespace. I added a comment in the docs that you shouldn't disable the root namespace when using nested namespaces.

**Comment by Frando** [comment]
On a related note - I'm not usually using namespaces, but ES modules, which seems to be the general recommendation for new typescript projects. It would be nice to support this fully in `typescript-type-def`. Currently, I'm exporting different sets of types to different files, and manually import them. This is not ideal, because the basic types (e.g. `Usize`) would be emitted multiple times. Instead I could imagine an `#[type_def(module="foo")]` and `#[type_def(module="bar")]` attribute. A new function `write_definition_modules` would then work the same as currently, but returning a `HashMap<String,Vec<u8>>` of `filename => ts code`. E.g.:

```rust
#[derive(TypeDef)]
#[type_def(module = "foo.ts")]
struct Foo { a: usize }


#[derive(TypeDef)]
#[type_def(module = "bar.ts")]
struct Bar { b: usize, foo: Foo }

let mut options = DefinitionFileOptions::default();
options.root_namespace = None;
options.common_module = Some("common");

let modules = write_definition_modules::<_(Foo,Bar)>(options)?;
for (name, code) in modules.iter() {
    std::fs::write(name, code).unwrap();
}
```
which would lead to:
```typescript
// base.ts
type Usize=number;
// foo.ts
import * as base from "./base";
export type Foo={"a":base.Usize;};
// bar.ts
import * as base from "./base";
import * as foo from "./foo";
export type Foo={"a":base.Usize;"foo":foo.Foo};
```

File: src/emit.rs [hunk 2]
```diff
@@ -502,7 +504,7 @@ impl Default for DefinitionFileOptions<'_> {
 /// Note that the TypeScript code generated by this library is not very
 /// human-readable. To make the code human-readable, use a TypeScript code
 /// formatter (such as [Prettier](https://prettier.io/)) on the output.
-pub fn write_definition_file<W, T: ?Sized>(
+pub fn write_definition_file<W, T>(
```
[the remark this answers is not in the document]
Comment (Frando) [reply src/emit.rs]: This was a leftover when trying to fix inbetween compiler errors I had, but it is not needed. Removed it.

File: src/impls.rs (hunk 1 above)
Comment (Frando) [reply src/impls.rs]: Yup, not needed & removed. Thanks!

**Comment by dbeckwith** [comment]
The modules idea is interesting, but my initial reaction is that it seems like it would add a lot of complexity to this library which I want to avoid unless it's a common use case. Could you open a separate issue to discuss that idea so I can get a clear sense of your use case for it?
